### PR TITLE
GPU integration into amps mpi1 layer (+minor updates)

### PR DESCRIPTION
**Pull request by hokkanen**
This PR integrates GPU capabilities into mpi1 amps layer with a cleaner implementation and better performance and error handling over the earlier amps cuda layer. The GPU packing functions are contained in a separate amps_gpupacking.cu file. By default, GPU packing is used, but CPU pointers are passed to the MPI library such that CUDA-aware MPI library is not required. A runtime environment variable `PARFLOW_USE_GPUDIRECT=1` disables host staging and enables GPU pointers to be passed to MPI library improving performance (but requires a CUDA-aware MPI library). Amps cuda layer becomes deprecated and may be removed in the future to reduce code duplication and maintenance burden.

## List of changes:
- Integrate GPU capabilities into mpi1 layer
- Fix amps smpi layer compile issues
- Fix amps smpi layer memory leak (persistent MPI_Requests not freed)
- Fix amps cuda layer memory leak (persistent MPI_Requests not freed)
- Disable unified memory use in clustering.c to improve performance
- Add CUDA 11 support
- Update CUDA documentation

**Includes CLM PR** [**#272**](https://github.com/parflow/parflow/pull/272)

### CODE REVIEW

File: pfsimulator/amps/mpi1/amps_exchange.c [hunk 1]
```diff
@@ -28,11 +28,126 @@
...
+                   i, &combuf, &size);
+      // if(errchk) printf("GPU unpacking failed at line: %d\n", errchk);
+    }
+    for (i = 0; i < handle->package->num_recv; i++)
+    {
+      amps_gpu_sync_streams(i);
+      AMPS_CLEAR_INVOICE(handle->package->recv_invoices[i]);
+    }
+    free(status);
+    (void)errchk;
```
Comment (smithsg84) [inline pfsimulator/amps/mpi1/amps_exchange.c:62]: remove errorchk if not using

File: pfsimulator/amps/mpi1/amps_exchange.c [hunk 2]
```diff
@@ -28,11 +28,126 @@
...
+                 handle->package->recv_invoices[i], 
+                   i, &combuf, &size);
+      // if(errchk) printf("GPU unpacking failed at line: %d\n", errchk);
+    }
+    for (i = 0; i < handle->package->num_recv; i++)
+    {
+      amps_gpu_sync_streams(i);
+      AMPS_CLEAR_INVOICE(handle->package->recv_invoices[i]);
+    }
+    free(status);
```
Comment (smithsg84) [inline pfsimulator/amps/mpi1/amps_exchange.c:61]: This should be moved up to the if clause.   If status is not NULL initialized I think this is a potential bug.

File: pfsimulator/clm/clm.F90 [hunk 3]
```diff
@@ -315,7 +315,7 @@ subroutine clm_lsm(pressure,saturation,evap_trans,topo,porosity,pf_dz_mult,istep
      !=== Initialize CLM and DIAG variables
      if (clm_write_logs==1) write(999,*) "Initialize CLM and DIAG variables"
      do t=1,drv%nch 
-        clm%kpatch = t
+        clm(t)%kpatch = t
```
Comment (smithsg84) [inline pfsimulator/clm/clm.F90:318]: Commits should be targeted at a single issues.   Either CLM patch needs to be removed from the commit or this won't be merged until the CLM PR has been approved.

File: pfsimulator/amps/mpi1/amps_gpupacking.cu [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+  .stream_id = NULL,
+  .num_streams = 0,
+  .reqs_since_sync = 0
+};
+
+/* Use page-locked host staging buffer by default */
+static int ENFORCE_HOST_STAGING = 1;
+
+/* Check for GPUDirect environment variable */
+void amps_gpu_check_env(){
```
Comment (smithsg84) [inline pfsimulator/amps/mpi1/amps_gpupacking.cu:76]: Functions should be documented with doxygen style comments.

**Review by smithsg84 — CHANGES_REQUESTED** [review]
Please add a bit more documentation, I'm trying to fix the lack of code documentation in new code so don't follow the existing horrible pattern.

File: pfsimulator/clm/clm.F90 (hunk 3 above)
Comment (smithsg84) [reply pfsimulator/clm/clm.F90]: CLM Patch was aproved.  This was fixed in #272

File: pfsimulator/amps/mpi1/amps_exchange.c (hunk 2 above)
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps_exchange.c:61]: Could you clarify this a bit more, maybe give line numbers what should be moved and where?

File: pfsimulator/amps/mpi1/amps.h [hunk 5]
```diff
@@ -1049,7 +1025,24 @@ void amps_ReadDouble(amps_File file, double *ptr, int len);
 #define amps_Error(name, type, comment, operation) \
   printf("%s : %s\n", name, comment)
 
-#ifdef __CUDACC__
+#ifdef PARFLOW_HAVE_CUDA
+#include <cuda.h>
```
[the remark this answers is not in the document]
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps.h]: Done

File: pfsimulator/amps/mpi1/amps.h [hunk 6]
```diff
@@ -1049,7 +1025,24 @@ void amps_ReadDouble(amps_File file, double *ptr, int len);
...
+#include <cuda_runtime.h>
+
+/*--------------------------------------------------------------------------
+ * Amps defines with CUDA
+ *--------------------------------------------------------------------------*/
+#define AMPS_GETRBUF 1
```
[the remark this answers is not in the document]
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps.h:1048]: Done

File: pfsimulator/amps/mpi1/amps.h [hunk 7]
```diff
@@ -1087,6 +1080,23 @@ static inline void amps_rmm_error(rmmError_t err, const char *file, int line) {
 }
 #endif
 
+/*--------------------------------------------------------------------------
+ * Amps gpu structs for global amps variables
+ *--------------------------------------------------------------------------*/
+typedef struct _amps_GpuBuffer {
```
[the remark this answers is not in the document]
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps.h]: Done

File: pfsimulator/amps/mpi1/amps.h [hunk 8]
```diff
@@ -1167,8 +1177,79 @@ static inline void _amps_tfree_cuda(void *ptr)
 /** Same as \ref amps_TFree but deallocates managed memory */
 #define amps_TFree_managed(ptr) if (ptr) _amps_tfree_cuda(ptr); else {}
 
+/*--------------------------------------------------------------------------
+ * Define amps GPU kernels
+ *--------------------------------------------------------------------------*/
+
+#ifdef __CUDACC__
+extern "C++"{
+template <typename T>
```
[the remark this answers is not in the document]
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps.h]: Done

File: pfsimulator/amps/mpi1/amps_exchange.c [hunk 9]
```diff
@@ -28,11 +28,126 @@
...
+                                  handle->package->num_send), sizeof(MPI_Status));
+
+    MPI_Waitall(handle->package->num_recv + handle->package->num_send,
+                handle->package->recv_requests, status);
+    for (i = 0; i < handle->package->num_recv; i++)
+    {
+      errchk = amps_gpupacking(AMPS_UNPACK, 
+                 handle->package->recv_invoices[i], 
+                   i, &combuf, &size);
+      // if(errchk) printf("GPU unpacking failed at line: %d\n", errchk);
```
[the remark this answers is not in the document]
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps_exchange.c]: Done

File: pfsimulator/amps/mpi1/amps_exchange.c (hunk 1 above)
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps_exchange.c]: Done

File: pfsimulator/amps/mpi1/amps_exchange.c (hunk 2 above)
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps_exchange.c]: I am still confused about this, but I removed the recurring alloc/dealloc, and the alloc/dealloc is now done when calling amps_NewPackage/amps_FreePackage.

File: pfsimulator/amps/mpi1/amps_gpupacking.cu (hunk 4 above)
Comment (hokkanen) [reply pfsimulator/amps/mpi1/amps_gpupacking.cu:169]: Done

File: pfsimulator/amps/mpi1/amps_exchange.c (hunk 2 above)
Comment (smithsg84) [reply pfsimulator/amps/mpi1/amps_exchange.c]: OK I got messed up in scoping, thought the free(status) was outside the L42 if.

**Review by smithsg84 — APPROVED** [review]

**Comment by smithsg84** [comment]
Thanks for updating PR.

**Pull request merged**